File: articles/developmental-systems-theory-and-the-ecology-of-human-growth/python/model_developmental_systems.py
#!/usr/bin/env python3
"""Model synthetic developmental systems theory and ecological development."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "developmental_systems_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_developmental_systems_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    development_model = smf.ols(
        """
        development_score ~ lag_score + time + biological_sensitivity +
        current_family + current_peer + school_climate +
        curriculum_opportunity + neighborhood_safety + service_access +
        material_security + intervention_exposure + ecological_stress +
        biological_sensitivity:current_family +
        biological_sensitivity:ecological_stress
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    stress_model = smf.ols(
        """
        ecological_stress ~ time + current_family + current_peer +
        school_climate + curriculum_opportunity + neighborhood_safety +
        service_access + material_security + intervention_exposure
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_developmental_systems_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(development_model.summary().as_text())
        f.write("\n\nECOLOGICAL STRESS MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(stress_model.summary().as_text())

    trajectory = panel.groupby(["time", "intervention_exposure"], as_index=False).agg(
        average_development=("development_score", "mean"),
        average_ecological_support=("ecological_support", "mean"),
        average_ecological_stress=("ecological_stress", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory["group"] = trajectory["intervention_exposure"].map({0: "No intervention", 1: "Intervention"})
    trajectory.to_csv(OUTPUTS_DIR / "python_development_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("group"):
        plt.plot(group["time"], group["average_development"], marker="o", label=group_name)
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Developmental Systems Trajectories")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_development_trajectory.png", dpi=160)
    plt.close()

    support_trajectory = panel.groupby("time", as_index=False).agg(
        average_ecological_support=("ecological_support", "mean"),
        average_ecological_stress=("ecological_stress", "mean"),
        average_family=("current_family", "mean"),
        average_peer=("current_peer", "mean"),
    )
    support_trajectory.to_csv(OUTPUTS_DIR / "python_ecological_support_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(support_trajectory["time"], support_trajectory["average_ecological_support"], marker="o", label="ecological support")
    plt.plot(support_trajectory["time"], support_trajectory["average_ecological_stress"], marker="o", label="ecological stress")
    plt.xlabel("Time")
    plt.ylabel("Average ecological index")
    plt.title("Synthetic Ecological Support and Stress")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_ecological_support_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("ecological_support_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_development=("development_score", "mean"),
        average_ecological_support=("ecological_support", "mean"),
        average_ecological_stress=("ecological_stress", "mean"),
        average_family=("current_family", "mean"),
        average_peer=("current_peer", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_ecological_support_profiles.csv", index=False)

    print("Wrote Python developmental systems model outputs.")


if __name__ == "__main__":
    main()

File: articles/developmental-systems-theory-and-the-ecology-of-human-growth/r/model_developmental_systems.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "developmental_systems_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_developmental_systems_panel.py first.")

panel <- read.csv(data_path)

development_model <- lmer(
  development_score ~ time + biological_sensitivity +
    current_family + current_peer + school_climate +
    curriculum_opportunity + neighborhood_safety + service_access +
    material_security + intervention_exposure + ecological_stress +
    biological_sensitivity:current_family +
    biological_sensitivity:ecological_stress +
    (1 + time | school_id/child_id),
  data = panel
)

stress_model <- lmer(
  ecological_stress ~ time + current_family + current_peer +
    school_climate + curriculum_opportunity + neighborhood_safety +
    service_access + material_security + intervention_exposure +
    (1 + time | school_id/child_id),
  data = panel
)

capture.output(
  list(
    development_model = summary(development_model),
    ecological_stress_model = summary(stress_model)
  ),
  file = file.path(outputs_dir, "r_developmental_systems_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, intervention_exposure) |>
  summarize(
    average_development = mean(development_score),
    average_ecological_support = mean(ecological_support),
    average_ecological_stress = mean(ecological_stress),
    standard_error = sd(development_score) / sqrt(n()),
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    .groups = "drop"
  ) |>
  mutate(group = ifelse(intervention_exposure == 1, "Intervention", "No intervention"))

write.csv(trajectory, file.path(outputs_dir, "r_development_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_development_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development, linetype = group)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = group), alpha = 0.12) +
    labs(title = "Synthetic Developmental Systems Trajectories", x = "Time", y = "Average development", linetype = "Group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

support <- panel |>
  group_by(time) |>
  summarize(
    average_ecological_support = mean(ecological_support),
    average_ecological_stress = mean(ecological_stress),
    average_family = mean(current_family),
    average_peer = mean(current_peer),
    .groups = "drop"
  )

write.csv(support, file.path(outputs_dir, "r_ecological_support_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_ecological_support_trajectory.png"),
  plot = ggplot(support, aes(x = time)) +
    geom_line(aes(y = average_ecological_support, linetype = "ecological support"), linewidth = 1) +
    geom_line(aes(y = average_ecological_stress, linetype = "ecological stress"), linewidth = 1) +
    labs(title = "Synthetic Ecological Support and Stress", x = "Time", y = "Average ecological index", linetype = "Index") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R developmental systems outputs.")
